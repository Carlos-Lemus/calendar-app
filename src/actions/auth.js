import Swal from "sweetalert2";
import { fetchWithoutToken, fetchWithToken } from "../helpers/fetch";
import { types } from "../types/types";

export const startLogin = (email, password) => {
    return async(dispatch) => {
        
        const req = await fetchWithoutToken("auth", "POST", {email, password});
        const body = await req.json();

        if(body.ok) {

            localStorage.setItem("token", body.token);
            localStorage.setItem("token-init-date", new Date().getTime());

            dispatch(login({
                uid: body.uid,
                name: body.name
            }));
        } else {
            Swal.fire("Error", body.msg, "error");
        }

    };
};

export const startRegister = (name, email, password) => {
    return async(dispatch) => {
        const resp = await fetchWithoutToken("auth/new", "POST", {name, email, password});
        const body = await resp.json();
        
        if(body.ok) {
            localStorage.setItem("token", body.token);
            localStorage.setItem("token-init-date", new Date().getTime());
            
            dispatch(login( { name: body.name, uid: body.uid } ) );
            
        } else {
            Swal.fire("Error", body.msg, "error");
        }
    };
};

export const startChecking = () => {
    return async(dispatch) => {
        const resp = await fetchWithToken("auth/renew");
        const body = await resp.json();
        
        if(body.ok) {
            localStorage.setItem("token", body.token);
            localStorage.setItem("token-init-date", new Date().getTime());
            
            dispatch(login( { name: body.name, uid: body.uid } ) );
            
        } else {
            dispatch(finishChecking());
        }
    };
};

export const startLogout = () => {
    return async(dispatch) => {

        localStorage.clear();

        dispatch(logout());
    }
};

const logout = () => ({ type: types.authLogout });

const login = ( user ) => ({
    type: types.authLogin,
    payload: user
});

export const finishChecking = () => ({
    type: types.authCheckingFinish
});

