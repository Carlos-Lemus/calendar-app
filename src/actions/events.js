import Swal from "sweetalert2";
import { fetchWithToken } from "../helpers/fetch";
import { prepareEvents } from "../helpers/prepareEvents";
import { types } from "../types/types";

export const startLoadingEvents = () => {
    return async(dispatch) => {
        try {
            const res = await fetchWithToken("events/");
            const body = await res.json();
            
            const events = prepareEvents(body.events);

            dispatch(loadedEvents(events));
            
        } catch (error) {
            console.log(error)
        }
    }
};

const loadedEvents = (events) => ({
    type: types.eventLoaded,
    payload: events
});

export const startSetActiveEvent = (event) => ({
    type: types.eventSetActive,
    payload: event
});

export const startNewAddEvent = (event) => {
    return async (dispatch, getState) => {
        try {
            const res = await fetchWithToken("events", "POST", event);
            const body = await res.json();

            if (body.ok) {

                const { uid, name } = getState().auth;

                event.id = body.event.id;
                event.user = {
                    _id: uid,
                    name
                };

                dispatch(newAddEvent(event));
            } else {
                Swal.fire("Error", body.msg, "error");
            }
        } catch (error) {
            console.log(error)
        }
    };
};

const newAddEvent = (event) => ({
    type: types.eventAddNew,
    payload: event
});

export const startClearActiveEvent = () => ({
    type: types.eventClearActiveEvent
});

export const startUpdateEvent = (event) => {
    return async(dispatch) => {
        try {
            const res = await fetchWithToken(`events/${ event.id }`, "PUT", event);
            const body = await res.json();
            
            if(body.ok) {
                dispatch(updateEvent(event));
            } else {
                Swal.fire("Error", body.msg, "error");
            }

        } catch (error) {
            console.log(error);
        }
    };
};

const updateEvent = (event) => ({
    type: types.eventUpdate,
    payload: event
});

export const startDeleteEvent = () => {

    return async(dispatch, getState) => {

        try {
            const {id} = getState().calendar.activeEvent;

            const res = await fetchWithToken(`events/${id}`, "DELETE");
            const body = await res.json();

            if(body.ok) {

                dispatch(deleteEvent());       
                dispatch(startClearActiveEvent());

            } else {
                Swal.fire("Error", body.msg, "error");
            }

        } catch (error) {
            console.log(error)
        }
    };

};

const deleteEvent = () => ({
    type: types.eventDeleted
});