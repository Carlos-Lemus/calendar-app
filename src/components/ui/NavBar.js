import React from 'react'

export const NavBar = () => {
    return (
        <nav className="navbar navbar-dark bg-dark mb-4">
            <span className="navbar-brand ">
                Carlos
            </span>

            <button className="btn btn-outline-danger">
                <i className="fas fa-sign-out-alt"></i>
                <span className="ml-2">Salir</span>
            </button>
            
        </nav>
    )
}
