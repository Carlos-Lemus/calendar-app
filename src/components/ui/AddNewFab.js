import React from 'react'
import { useDispatch } from 'react-redux';
import { startOpenModal } from '../../actions/ui';

export const AddNewFab = () => {

    const dispatch = useDispatch()

    const handleClick = () => {
        dispatch(startOpenModal());
    }

    return (
        <button
            className="btn btn-primary fab"
            type="button"
            onClick={ handleClick }
        >
            <i className="fas fa-plus"></i>
        </button>
    )
}
