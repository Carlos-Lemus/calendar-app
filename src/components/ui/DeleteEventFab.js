import React from 'react'
import { useDispatch } from 'react-redux'
import { startDeleteEvent } from '../../actions/events'

export const DeleteEventFab = () => {

    const dispatch = useDispatch()

    const handleClick = () => {
        dispatch(startDeleteEvent());
    }

    return (
        <button 
            onClick={ handleClick }
            className="btn btn-danger fab-danger"
            type="button"
        >
            <i className="fas fa-trash"></i>
            <span>Borrar evento</span>
        </button>
    )
}
