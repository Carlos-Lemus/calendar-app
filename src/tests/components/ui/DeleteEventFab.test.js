import configMockStore from "redux-mock-store";
import thunk from "redux-thunk";
import "@testing-library/jest-dom";
import { mount } from "enzyme";
import { Provider } from "react-redux";
import { DeleteEventFab } from "../../../components/ui/DeleteEventFab";
import { startDeleteEvent } from "../../../actions/events";

const middlewares = [thunk];
const mockStore = configMockStore(middlewares);
Storage.prototype.setItem = jest.fn();

const initialState = {};

let store = mockStore(initialState);
store.dispatch = jest.fn();

jest.mock("../../../actions/events", () => ({
    startDeleteEvent: jest.fn()
}));

const wrapper = mount(
    <Provider store={store}>
        <DeleteEventFab />
    </Provider>
);

describe("Pruebas en <DeleteEventFab />", () => {

    test('Haciendo snapshot de <DeleteEventFab />', () => {
        expect(wrapper).toMatchSnapshot();
    });
    
    test('debe ejecutar el evento handleClick al hacer click', () => {
        wrapper.find("button").prop("onClick")();

        expect(store.dispatch).toHaveBeenCalledTimes(1);
        expect(startDeleteEvent).toHaveBeenCalledTimes(1);
    })
    

});