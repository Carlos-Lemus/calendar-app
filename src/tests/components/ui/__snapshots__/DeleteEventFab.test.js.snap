// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <DeleteEventFab /> Haciendo snapshot de <DeleteEventFab /> 1`] = `
<button
  className="btn btn-danger fab-danger"
  onClick={[Function]}
  type="button"
>
  <i
    className="fas fa-trash"
  />
  <span>
    Borrar evento
  </span>
</button>
`;
